pkg/scheduler/algorithm/predicates/csi_volume_predicate.go

```diff
@@ -51,60 +65,67 @@ func NewCSIMaxVolumeLimitPredicate(
 		return true, nil, nil
 	}
+
 	// If a pod doesn't have any volume attached to it, the predicate will always be true.
-	// Thus we make a fast path for it, to avoid unnecessary computations in this case.
+	// Thus we make a fast path for it to avoid unnecessary computations in this case.
 	if len(pod.Spec.Volumes) == 0 {
 		return true, nil, nil
 	}
 
-	nodeVolumeLimits := nodeInfo.VolumeLimits()
-
-	// if node does not have volume limits this predicate should exit
-	if len(nodeVolumeLimits) == 0 {
-		return true, nil, nil
-	}
-
-	// a map of unique volume name/csi volume handle and volume limit key
-	newVolumes := make(map[string]string)
-	if err := c.filterAttachableVolumes(pod.Spec.Volumes, pod.Namespace, newVolumes); err != nil {
+	// A map of unique volume name/csi volume handle and volumeIdentity
+	newVolumes := make(map[string]volumeIdentity)
+	if err := c.filterAttachableVolumes(nodeInfo, pod.Spec.Volumes, pod.Namespace, newVolumes); err != nil {
 		return false, nil, err
 	}
 
 	if len(newVolumes) == 0 {
 		return true, nil, nil
 	}
 
-	// a map of unique volume name/csi volume handle and volume limit key
-	attachedVolumes := make(map[string]string)
+	// A map of unique volume name/csi volume handle and volumeIdentity
+	attachedVolumes := make(map[string]volumeIdentity)
 	for _, existingPod := range nodeInfo.Pods() {
-		if err := c.filterAttachableVolumes(existingPod.Spec.Volumes, existingPod.Namespace, attachedVolumes); err != nil {
+		if err := c.filterAttachableVolumes(nodeInfo, existingPod.Spec.Volumes, existingPod.Namespace, attachedVolumes); err != nil {
 			return false, nil, err
 		}
 	}
 
-	newVolumeCount := map[string]int{}
 	attachedVolumeCount := map[string]int{}
-
-	for volumeName, volumeLimitKey := range attachedVolumes {
-		if _, ok := newVolumes[volumeName]; ok {
-			delete(newVolumes, volumeName)
+	for volumeUniqueName, vIdentity := range attachedVolumes {
+		if _, ok := newVolumes[volumeUniqueName]; ok {
+			delete(newVolumes, volumeUniqueName)
 		}
-		attachedVolumeCount[volumeLimitKey]++
+		attachedVolumeCount[vIdentity.volumeLimitKey]++
+	}
+	type volumeTypeCount struct {
+		count      int
+		volumeType string
 	}
 
-	for _, volumeLimitKey := range newVolumes {
-		newVolumeCount[volumeLimitKey]++
+	newVolumeCount := map[string]volumeTypeCount{}
+	for _, vIdentity := range newVolumes {
+		vt, ok := newVolumeCount[vIdentity.volumeLimitKey]
+		if !ok {
+			vt = volumeTypeCount{count: 0, volumeType: vIdentity.volumeType}
+		}
+		vt.count++
+		newVolumeCount[vIdentity.volumeLimitKey] = vt
 	}
 
-	for volumeLimitKey, count := range newVolumeCount {
+	// We don't return early if the node has no limits because we want to
+	// report an error if a CSI driver is not installed in the node.
+	nodeVolumeLimits := nodeInfo.VolumeLimits()
+	for volumeLimitKey, vt := range newVolumeCount {
 		maxVolumeLimit, ok := nodeVolumeLimits[v1.ResourceName(volumeLimitKey)]
-		if ok {
+		if !ok && vt.volumeType == csiVolumeType {
```

Comment:
@bertinatto @msau42 So I made this code so that, if we know that volume is really of CSI type then and only then we prevent pod scheduling. For other cases we do the regular stuff. Also added unit tests for bunch of scenarios from here.

Reply:
> if we know that volume is really of CSI type

We still can't tell if the volume is CSI unless the translation library knows it. 

So the predicate will stop scheduling pods with unbound volumes (i.e., when the SC has `volumeBindingMode: WaitForFirstConsumer`) on nodes where the driver is not installed, **if** the volume belongs to a group of CSI drivers (EBS, GCEPD, etc.) 

However, the behaviour will be different for all other CSI drivers: k8s will keep scheduling pods on nodes where the driver is not installed.

IMO we should have a consistent behaviour among all CSI drivers, and keep the current behaviour at least until we can reliable tell whether an unbound volume is CSI or not.

Reply:
An option to solve this would be to have the predicate:

1) Look at the SC provisioner name (if the volume is unbound)
2) Then check if there's a `CSIDriver` object with the same name

We'd have to cache another object, but that give us confidence that the volume is really CSI. I discussed this with @jsafrane and he mentioned that `CSIDriver` object is currently optional, so we'd have to make it non-optional as well.

Reply:
Yeah - we can use `CSIDriver` object but since it will require informer etc, that can be added as a different PR.

>However, the behaviour will be different for all other CSI drivers: k8s will keep scheduling pods on nodes where the driver is not installed.

This applies only when a PVC is being bound for the first time, in all other cases when PVC is already bound we will correctly stop pods from scheduling if driver is not installed. 

> IMO we should have a consistent behaviour among all CSI drivers, and keep the current behaviour at least until we can reliable tell whether an unbound volume is CSI or not.

Even if we use `CSIDriver` object the proposed feature (of preventing pod scheduling) will only work when `CSIDriver` object is installed. So it will always be a opt-in. 

We have two options:

1. Remove new behaviour entirely, because not counting PVCs if they are unbound is not an option. It will result in regression.
2. Keep counting unbound PVCs and only prevent pod scheduling (for no driver case) when we know for sure that PVC is of CSI type. 

We can extend#2 with `CSIDriver` check in future PR. I prefer option#2 rather than dropping 
"new behaviour" altogether. I am not sure about consistency argument - scheduler does its best to prevent scheduling whenever it can correctly determine PVC is of `CSI` type.

Reply:
> I am not sure about consistency argument - scheduler does its best to prevent scheduling whenever it can correctly determine PVC is of CSI type.

I used the consistency argument because given the exact same conditions, k8s will act differently depending on the CSI driver.

Another thing I was thinking is that a CSI driver author will have little to no power to change that, unless their driver is a target to an in-tree migration. The `CSIDriver` solution could be helpful in these cases.

IDK if my vote counts, but I'd keep this PR only focused on the volume limits and handle the "new behaviour" in another PR, where we can implement the `CSIDriver` look up only. 

That way, we'll be able to document that in order to prevent pods (with unbound PVCs) from scheduling in nodes where the driver is not installed, driver authors should make sure that a `CSIDriver` object is properly created for their driver.

Reply:
> IDK if my vote counts, but I'd keep this PR only focused on the volume limits and handle the "new behaviour" in another PR, where we can implement the CSIDriver look up only.

I vote for this too. Not sure if we need an extra predicate (because it would loop through all volumes in a pod and dereference PVCs again, as this predicate does), but it should be definitely a separate PR.

Reply:
okay so just so I understand this. Even if we know volume being considered is of CSI type because it has PV source present, we are going to require `CSIDriver` object to prevent pod scheduling?  That is to ensure we use same thing as source of truth?

Reply:
For PVC bound case, we can still use PV.CSISource.  For PVC unbound case, we can use CSIDriver object.  I'm fine with this behavior being opt-in.  Scheduling based on CSI driver node installation is a new feature and so far I'm not aware of any plugins that are going to be installed on only a subset of nodes.

Reply:
I don't have strong opinions on whether or not it should be a separate PR or not.  Mostly depends on your bandwidth/confidence in implementing this by early 1.16 so we can get enough scalability soak time.  It would help make this already big PR easier to review potentially.

Reply:
Alright I will leave this with @bertinatto . If you want to remove new behavior of preventing pod scheduling if driver is not installed on the node then feel free to delete all the code and tests related to it.

We will try and cover this in a different PR.